pkg/resource/deploy/step_generator.go

```diff
@@ -262,6 +279,9 @@ func (sg *stepGenerator) GenerateReadSteps(event ReadResourceEvent) ([]Step, err
 // If the given resource is a custom resource, the step generator will invoke Diff and Check on the
 // provider associated with that resource. If those fail, an error is returned.
 func (sg *stepGenerator) GenerateSteps(event RegisterResourceEvent) ([]Step, error) {
+	sg.urnLock.lock()
+	defer sg.urnLock.unlock()
```

Comment:
this is a pretty big lock! I think that's probably fine, but I admit that it makes me a little nervous w.r.t. reasoning about when the lock is/isn't held. I think we'd benefit from an expanded commit message that explains the locking strategy + its reasoning in detail.

Reply:
Yeah, I tried finer grained approaches but could never be sure that there wouldn't be a lock ordering issue. I'll expand the commit comment _per_ your suggestion, the fundamental reasoning is that the slow step is provider Check or Diff and that this unlocks for those operations

Reply:
bumping this--still looking for a detailed comment that explains the locking / how we avoid deadlock / etc. I'd specifically like to see docs on when per-resource locks are taken, how many are held in aggregate, etc.

Reply:
Ugh. Sorry, I wrote a long and detailed commit description for this, but must have not saved it. Now I have to remember what I wrote...